From acbc93b71fc35035f3a2fe9615f6ea0b37abff2f Mon Sep 17 00:00:00 2001
From: Mohammed Javith Akthar M <mohammedjavitham@ami.com>
Date: Fri, 4 Oct 2024 07:43:32 -0400
Subject: [PATCH 11/11] OT-7707: Fix Command Injection in Virtual Media

Root Cause:
The NFS implementation uses the `system()` API for executing mount
operations. Due to the lack of strict input sanitization in the upper
layers (e.g., WebUI/Redfish), this makes NFS redirection vulnerable
to command injection attacks.

Fix:
Replaced the `system()` API with the `mount()` API for NFS
implementation to address this vulnerability on the server side.
Strict input sanitization will also be added in the upper layers in
a separate commit.

Signed-off-by: Mohammed Javith Akthar M <mohammedjavitham@ami.com>
---
 src/nfs.hpp | 119 ++++++++++++++++++++++++++++++----------------------
 1 file changed, 69 insertions(+), 50 deletions(-)

diff --git a/src/nfs.hpp b/src/nfs.hpp
index e9e7435..83243a1 100644
--- a/src/nfs.hpp
+++ b/src/nfs.hpp
@@ -9,67 +9,86 @@
 #include <optional>
 
 namespace fs = std::filesystem;
-#define MAX_CMD_NAME_LEN                ((4*256) + 256)
-#define MAX_IMG_PATH 256
-
-static const char* chkmountprocess = "pid=$!;"
-                                     "a=1;"
-                                     "while [ $a -lt 10 ];"
-                                     "do sleep 1;"
-                                     "a=`expr $a + 1`;"
-                                     "if [ -f /proc/$pid/exe ];"
-                                     "then continue;"
-                                     "else break;"
-                                     "fi;"
-                                     "done;"
-                                     "if [ $a == 10 ];"
-                                     "then echo killed;"
-                                     "kill -9 $pid;"
-                                     "fi;";
 
 class NfsShare
 {
-    public:
-        NfsShare(const fs::path& mountDir) : mountDir(mountDir)
+  public:
+    NfsShare(const fs::path& mountDir) : mountDir(mountDir)
     {
     }
 
-        bool mount(const fs::path& remote, bool rw)
+    bool mount(const fs::path& remote, bool rw)
+    {
+        LogMsg(Logger::Debug, "Trying to mount NFS remotely : ", remote);
+
+        const std::string params = "";
+        const std::string perm = rw ? "rw" : "ro";
+        std::string options = params + "," + perm;
+
+        std::string versionOpt = "nfs4";
+
+        auto ec = mountWithNfsVers(remote, options, versionOpt);
+
+        if (ec)
         {
-            LogMsg(Logger::Debug, "Trying to mount NFS remotely : ", remote);
-
-            char MountCmd [MAX_CMD_NAME_LEN]={0};
-            int size=sizeof(MountCmd);
-            char script[MAX_IMG_PATH*3] ={0};
-
-            int ret = snprintf (MountCmd, static_cast<size_t>(size), "mount -t nfs -o rw,nolock  %s %s",remote.c_str(), mountDir.c_str());
-            if(ret >= size || ret < 0)
-            {
-                std::cerr << "Buffer Overflow\n";
-                return false;
-            }
-
-            ret = snprintf(script,MAX_IMG_PATH *3," %s > /dev/null & %s",MountCmd,chkmountprocess);
-            if ((ret >= MAX_IMG_PATH*3) || (ret < 0))
-            {
-                std::cerr << "Buffer Overflow\n";
-                return false;
-            }
-
-            auto ec = system(script);
-            if( ec != 0)
-            {
-                LogMsg(Logger::Info, "Nfs Mount failed",
-                        " with ec = ", ec, " errno = ", errno);
-                return false;
-            }
+            versionOpt = "nfs";
+            ec = mountWithNfsVers(remote, options, versionOpt);
+        }
+
+        utils::secureCleanup(options);
 
+        if (ec)
+        {
+            return false;
+        }
+        return true;
+    }
+
+  private:
+    std::string mountDir;
+
+    bool parseNfsArgs(const std::string remote, std::string& remoteNfsPath,
+                      std::string& options)
+    {
+        std::string::size_type index = remote.find(":/", 0);
+        if ((index != std::string::npos) && (index > 0))
+        {
+            remoteNfsPath = remote.substr(index, remote.length() - 1);
+            options += (",nolock,addr=" + remote.substr(0, index));
             return true;
         }
+        return false;
+    }
 
-    private:
-        std::string mountDir;
+    int mountWithNfsVers(const fs::path& remote, std::string options,
+                         const std::string& version)
+    {
+        std::string remoteNfsPath = "";
 
-};
+        if (parseNfsArgs(remote.generic_string(), remoteNfsPath, options) !=
+            true)
+        {
+            LogMsg(Logger::Info, "NFS Mount failed when parsing remote path ",
+                   remote.generic_string());
+            return 1;
+        }
 
+        /* remote will have <remote_host>:<remote_path> format. Although it
+        ** works for SMB, it fails for NFS with parsing error.
+        ** So need to give in :<remotepath> and append addr=<remote_host> in
+        ** options to make it work */
+        auto ec = ::mount(remoteNfsPath.c_str(), mountDir.c_str(),
+                          version.c_str(), 0, options.c_str());
 
+        utils::secureCleanup(options);
+        utils::secureCleanup(remoteNfsPath);
+
+        if (ec)
+        {
+            LogMsg(Logger::Info, "Mount failed for ", version,
+                   " with ec = ", ec, " errno = ", errno);
+        }
+
+        return ec;
+    }
+};
-- 
2.43.0

